tests/Credentials/EcsCredentialProviderTest.php

```diff
@@ -311,4 +315,304 @@ private function getClientWithHeaderMiddleware($expectedValue)
+            }
+            return Promise\Create::promiseFor(
+                new Response(
+                    200,
+                    [],
+                    Psr7\Utils::streamFor(
+                        json_encode(call_user_func_array(
+                            [$this, 'getCredentialArray'],
+                            $creds
+                        ))
+                    )
+                )
+            );
+
+            return Promise\Create::rejectionFor([
+                'exception' => new \Exception(
+                    'Invalid request passed to test server'
+                )
+            ]);
+        };
+    }
+
+    /**
+     * @dataProvider failureTestCases
+     *
+     * @param $client
+     * @param \Exception $expected
+     */
+    public function testHandlesFailureScenarios($client, \Exception $expected)
+    {
+        $provider = new EcsCredentialProvider([
+            'client' => $client,
+            'retries' => 1,
+        ]);
+
+        try {
+            $provider()->wait();
+            $this->fail('Provider should have thrown an exception.');
+        } catch (\Exception $e) {
+            $this->assertInstanceOf(get_class($expected), $e);
+            $this->assertSame($expected->getMessage(), $e->getMessage());
+        }
+    }
+
+    public function failureTestCases()
+    {
+        $getRequest = new Psr7\Request('GET', '/latest');
+
+        $rejectionCreds = Promise\Create::rejectionFor([
+            'exception' => new RequestException(
+                '401 Unathorized',
+                $getRequest,
+                new Psr7\Response(401)
+            )
+        ]);
+        $connectException = new ConnectException(
+            'cURL error 28: Connection timed out after 1000 milliseconds',
+            new Psr7\Request('GET', '/latest')
+        );
+        $rejectionConnection = Promise\Create::rejectionFor([
+            'exception' => $connectException,
+        ]);
+
+        return [
+            'Non-retryable error' => [
+                $this->getTestClient(
+                    [
+                        $rejectionCreds,
+                    ]
+                ),
+                new CredentialsException(
+                    'Error retrieving credentials from container metadata after attempt 0/1 (401 Unathorized)'
+                )
+            ],
+            'Retryable error' => [
+                $this->getTestClient(
+                    [
+                        $rejectionConnection,
+                        $rejectionConnection,
+                    ]
+                ),
+                new CredentialsException(
+                    'Error retrieving credentials from container metadata after attempt 1/1 (cURL error 28: Connection timed out after 1000 milliseconds)'
+                )
+            ],
+        ];
+    }
+
+    public function testReadsRetriesFromEnvironment()
+    {
+        putenv('AWS_METADATA_SERVICE_NUM_ATTEMPTS=1');
```

Comment:
This needs to be cleaned up— it's being read by the `InstanceProfileProvider` tests 🙃